File: src/visualization/plot_discovery_velocity.py
import os
import json
import matplotlib.pyplot as plt

def main():
    report_file = os.path.join("reports", "summary", "sensitivity_analysis_report.json")
    if not os.path.exists(report_file):
        print(f"Report file not found: {report_file}")
        return

    with open(report_file, 'r', encoding='utf-8') as f:
        data = json.load(f)

    dv_summary = data.get("discovery_velocity_summary", {})
    if not dv_summary:
        print("No discovery_velocity_summary found.")
        return

    budgets = [int(b) for b in dv_summary.keys()]
    cumulative_K = [dv_summary[str(b)]["cumulative_failure_modes_K"] for b in budgets]
    velocities = [dv_summary[str(b)]["discovery_velocity_dK_dN"] for b in budgets]

    fig, ax1 = plt.subplots(figsize=(8, 5))

    color = 'tab:blue'
    ax1.set_xlabel('Sample Size (N iterations)', fontsize=12, fontweight='bold')
    ax1.set_ylabel('Cumulative Unique Failures Discovered K(N)', color=color, fontsize=12, fontweight='bold')
    ax1.plot(budgets, cumulative_K, marker='o', color=color, linewidth=2.5, label='K(N) Cumulative Failures')
    ax1.tick_params(axis='y', labelcolor=color)
    ax1.axvline(x=100000, color='red', linestyle='--', linewidth=1.5, label='100k Sampling Limit')

    ax2 = ax1.twinx()  
    color = 'tab:orange'
    ax2.set_ylabel('Discovery Velocity dK/dN', color=color, fontsize=12, fontweight='bold')
    ax2.plot(budgets, velocities, marker='s', color=color, linewidth=2, linestyle=':', label='dK/dN Velocity')
    ax2.tick_params(axis='y', labelcolor=color)

    plt.title('Monte Carlo Discovery Velocity & Failure Mode Saturation (dK/dN)', fontsize=13, fontweight='bold', pad=15)
    fig.tight_layout()

    out_dir = os.path.join("reports", "figures")
    os.makedirs(out_dir, exist_ok=True)
    out_img = os.path.join(out_dir, "discovery_velocity_plateau.png")
    plt.savefig(out_img, dpi=300)
    print(f"Discovery velocity plot saved to: {out_img}")

if __name__ == "__main__":
    main()

File: src/visualization/plot_discovery_velocity.R
# Discovery Velocity Publication Plotting Script in R (ggplot2)
library(jsonlite)
library(ggplot2)
library(scales)

# Read report
report_path <- "reports/summary/sensitivity_analysis_report.json"
data <- fromJSON(report_path)

dv_summary <- data$discovery_velocity_summary
budgets <- as.numeric(names(dv_summary))

df <- data.frame(
  N = budgets,
  K = sapply(dv_summary, function(x) x$cumulative_failure_modes_K),
  Velocity = sapply(dv_summary, function(x) x$discovery_velocity_dK_dN)
)

# Scaling factor for dual Y axis
coeff <- max(df$K) / max(df$Velocity)

p <- ggplot(df, aes(x = N)) +
  # Soft pastel area fill under K(N)
  geom_area(aes(y = K), fill = "#EBF4FF", alpha = 0.7) +
  
  # Saturation threshold vertical line at 100k
  geom_vline(xintercept = 100000, linetype = "dashed", color = "#C53030", linewidth = 0.8) +
  
  # Primary Line: Cumulative Failure Modes K(N)
  geom_line(aes(y = K, color = "Cumulative Failures K(N)"), linewidth = 1.3) +
  geom_point(aes(y = K, color = "Cumulative Failures K(N)"), size = 3, shape = 19) +
  
  # Secondary Line: Discovery Velocity dK/dN
  geom_line(aes(y = Velocity * coeff, color = "Discovery Velocity dK/dN"), linewidth = 1.1, linetype = "dotdash") +
  geom_point(aes(y = Velocity * coeff, color = "Discovery Velocity dK/dN"), size = 2.8, shape = 15) +
  
  # Axes scales and formatting
  scale_y_continuous(
    name = "Cumulative Unique Failures Discovered K(N)",
    limits = c(0, 45),
    expand = c(0, 0),
    sec.axis = sec_axis(~./coeff, name = "Discovery Velocity (dK/dN)")
  ) +
  scale_x_continuous(
    name = "Sample Budget (N Iterations)",
    labels = comma,
    breaks = c(0, 25000, 50000, 75000, 100000),
    limits = c(0, 105000),
    expand = c(0, 0)
  ) +
  
  # Color Palette
  scale_color_manual(
    values = c(
      "Cumulative Failures K(N)" = "#1A365D",
      "Discovery Velocity dK/dN" = "#C53030"
    )
  ) +
  
  # Clean Academic Journal Theme
  theme_bw(base_size = 13, base_family = "sans") +
  theme(
    plot.background = element_rect(fill = "#FFFFFF", color = NA),
    panel.background = element_rect(fill = "#FFFFFF", color = NA),
    panel.border = element_rect(color = "#2D3748", fill = NA, linewidth = 0.8),
    panel.grid.major = element_line(color = "#EDF2F7", linewidth = 0.5),
    panel.grid.minor = element_blank(),
    
    plot.title = element_text(face = "bold", size = 15, hjust = 0.5, color = "#1A202C", margin = margin(b = 12)),
    
    axis.title.x = element_text(face = "bold", color = "#2D3748", size = 12, margin = margin(t = 10)),
    axis.title.y.left = element_text(face = "bold", color = "#1A365D", size = 12, margin = margin(r = 10)),
    axis.title.y.right = element_text(face = "bold", color = "#C53030", size = 12, margin = margin(l = 10)),
    
    axis.text.x = element_text(color = "#4A5568", size = 10),
    axis.text.y.left = element_text(color = "#1A365D", size = 10, face = "bold"),
    axis.text.y.right = element_text(color = "#C53030", size = 10, face = "bold"),
    
    legend.position = "top",
    legend.title = element_blank(),
    legend.background = element_rect(fill = "#FFFFFF", color = "#E2E8F0", linewidth = 0.5),
    legend.key = element_rect(fill = "#FFFFFF", color = NA),
    legend.text = element_text(size = 11, face = "bold", color = "#2D3748"),
    legend.margin = margin(t = 4, b = 6, l = 10, r = 10)
  ) +
  labs(
    title = "Monte Carlo Discovery Velocity & Failure Mode Saturation (dK/dN)"
  )

# Save high-res publication PNG
ggsave("reports/figures/discovery_velocity_ggplot2.png", plot = p, width = 9, height = 5.5, dpi = 300, bg = "white")
cat("Saved publication-grade ggplot2 figure to reports/figures/discovery_velocity_ggplot2.png\n")
